pandas/tests/frame/methods/test_set_index.py

```diff
@@ -47,6 +47,18 @@ def test_set_index_empty_column(self):
         expected.index = MultiIndex.from_arrays([df["a"], df["x"]], names=["a", "x"])
         tm.assert_frame_equal(result, expected)
 
+    def test_set_index_empty_dataframe(self):
+        # GH#38419
+        df = DataFrame(
+            {"a": Series(dtype="datetime64[ns]"), "b": Series(dtype="int64"), "c": []}
+        )
+
+        if df.empty:
+            df = DataFrame(columns=list(df.columns.values))
+
+        expected = df.set_index(["a", "b"])
+        assert (df.loc[:, ["a", "b"]].dtypes == expected.index.to_frame().dtypes).all()
```

Comment:
This would work?

```python
def test_set_index_empty_dataframe(self):
        # GH#38419
        df = DataFrame(
            {"a": Series(dtype="datetime64[ns]"), "b": Series(dtype="int64"), "c": []}
        )

        df = DataFrame(columns=list(df.columns.values))
        expected = df.set_index(["a", "b"])
        tm.assert_frame_equal(df.loc[:, ['a', 'b']],
        expected.index.to_frame().reset_index(drop = True).set_index(d1.index))
```

Reply:
Sorry but absolutely not, this it too complicated 
Construct expected ahead of time and just do
```
tm.assert_frame_equal(result, expected)
```

Reply:
Make the construction as explicit as possible. The point here is to make it as easy as possible to see what the test does

Reply:
You want to do something like this 
``` python
In [21]: import pandas as pd
    ...: import pandas._testing as tm
    ...: 
    ...: df1 = pd.DataFrame({'a': pd.Series(dtype='datetime64[ns]'), 'b': pd.Series(dtype='int64'), 'c': []})
    ...: df2= df1.set_index(['a', 'b'])
    ...: 
    ...: result = df2.dtypes
    ...: expected = df1[['c']].dtypes # double square brackets to slice to a DataFrame not to a Series 
    ...: tm.assert_series_equal(result, expected)
```

Reply:
This solution would be valid?

```python
import pandas as pd
import pandas._testing as tm

df1 = pd.DataFrame({'a': pd.Series(dtype='datetime64[ns]'), 'b': pd.Series(dtype='int64'), 'c': []})
df2= df1.set_index(['a', 'b'])

result = df2.index.to_frame().dtypes
expected = df1[['a', 'b']].dtypes
tm.assert_series_equal(result, expected)
```